File: pkg/speechkit/stt/multilanguage_contract_test.go
package stt

import "testing"

// Mirrors the Kotlin gate at
// android/core/src/test/kotlin/io/kombify/speechkit/stt/MultilanguageContractTest.kt.
//
// SpeechKit does not narrow speech to one language. A pinned language breaks
// switching language mid-conversation and speaking different languages with
// different people in one session, and it is a silent data-loss bug: the same
// English audio returns a zero-length transcript when pinned to German, with
// HTTP 200 and no error. This has regressed repeatedly, so it is pinned by
// tests rather than by intent.
//
// Both directions are covered: an unset language must resolve to
// multilanguage, and a deliberate user pin must still survive — a gate that
// only forbade the value would break the supported override.

func TestIsMultilanguageAcceptsEveryUnpinnedSpelling(t *testing.T) {
	// The settings UI offers "auto" (Auto / provider detect) and "multi"
	// (Multilingual) as separate labels, so both have to resolve here.
	for _, language := range []string{"", "   ", "auto", "AUTO", "multi", "MULTI", " Multi "} {
		if !IsMultilanguage(language) {
			t.Errorf("IsMultilanguage(%q) = false, want true", language)
		}
	}
}

func TestIsMultilanguageRejectsALocale(t *testing.T) {
	for _, language := range []string{"de", "en", "de-DE", "pt-BR"} {
		if IsMultilanguage(language) {
			t.Errorf("IsMultilanguage(%q) = true, want false", language)
		}
	}
}

// The regression this fixes: "multi" was passed through verbatim because only
// "" and "auto" were normalized away. That is correct for Deepgram, whose API
// takes the literal token, and invalid for every provider that expects a
// language code or nothing at all — the same defect shape as the Android
// platform recognizer, where the literal token became an invalid BCP-47 tag.
func TestNormalizedRequestLanguageDoesNotForwardTheSentinel(t *testing.T) {
	for _, language := range []string{"", "auto", "multi"} {
		if got := normalizedRequestLanguage(language, false); got != "" {
			t.Errorf("normalizedRequestLanguage(%q) = %q, want \"\" so each adapter can express multilanguage natively", language, got)
		}
	}
}

func TestNormalizedRequestLanguageKeepsAnExplicitPin(t *testing.T) {
	if got := normalizedRequestLanguage("de", false); got != "de" {
		t.Errorf("normalizedRequestLanguage(\"de\") = %q, want \"de\" — an explicit pin is a supported override", got)
	}
	if got := normalizedRequestLanguage("pt-BR", false); got != "pt-BR" {
		t.Errorf("normalizedRequestLanguage(\"pt-BR\") = %q, want \"pt-BR\"", got)
	}
}

func TestDetectLanguageStillWins(t *testing.T) {
	if got := normalizedRequestLanguage("de", true); got != "" {
		t.Errorf("normalizedRequestLanguage(\"de\", detect=true) = %q, want \"\"", got)
	}
}

// The response direction, which the request-side gates above left open.
//
// Providers that cannot report a detected language used to fall back to a
// hardcoded "de" label while Deepgram and Google fell back to the sentinel.
// That inconsistency was not cosmetic: consumers pick a customization
// dictionary from Transcript.Language, so an unpinned huggingface or local
// session had German vocabulary applied to speech in any language.
//
// The rule is the same in both directions. A label is either something the
// provider actually reported, or the value that was explicitly asked for. It
// is never a locale nobody chose.
func TestReportedLanguageNeverInventsALocale(t *testing.T) {
	for _, tc := range []struct {
		name     string
		detected string
		asked    string
		want     string
	}{
		{name: "detected language wins", detected: "pt-BR", asked: "", want: "pt-BR"},
		{name: "detected wins over a pin", detected: "en", asked: "de", want: "en"},
		{name: "pin survives when nothing detected", detected: "", asked: "de", want: "de"},
		{name: "unpinned falls back to the sentinel", detected: "", asked: "", want: LanguageMulti},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if got := firstNonEmptyTrimmed(tc.detected, tc.asked, LanguageMulti); got != tc.want {
				t.Errorf("reported language = %q, want %q", got, tc.want)
			}
		})
	}
}

// APILanguage is what the adapters hand to firstNonEmptyTrimmed as the "asked
// for" value, so it must not leak the sentinel into a label either — reporting
// the literal "auto" would be as wrong as reporting "de".
func TestAPILanguageResolvesTheSentinelToEmpty(t *testing.T) {
	for _, language := range []string{"", "auto", "multi"} {
		resolved := ResolvedTranscribeOptions{Language: language}
		if got := resolved.APILanguage(); got != "" {
			t.Errorf("APILanguage(%q) = %q, want \"\" so the label falls through to the sentinel", language, got)
		}
	}
	resolved := ResolvedTranscribeOptions{Language: "de"}
	if got := resolved.APILanguage(); got != "de" {
		t.Errorf("APILanguage(\"de\") = %q, want \"de\"", got)
	}
}

File: android/core/src/test/kotlin/io/kombify/speechkit/stt/MultilanguageContractTest.kt
package io.kombify.speechkit.stt

import io.kombify.speechkit.stt.system.SystemSpeechRecognizerSession
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/**
 * The multilanguage contract, as a gate.
 *
 * SpeechKit does not narrow speech to one language: a pinned language breaks
 * switching language mid-conversation and speaking different languages with
 * different people in one session, and it is a silent data-loss bug (the same
 * English audio yields a zero-length transcript when pinned to German, with
 * HTTP 200 and no error).
 *
 * This has regressed repeatedly, so it is pinned by tests rather than by
 * intent. Both directions are covered: an unconfigured request must carry
 * multilanguage, and a deliberate user pin must still survive — a gate that
 * only forbade the field would break the supported override.
 */
class MultilanguageContractTest {

    @Test
    fun `an unconfigured request is multilanguage`() {
        assertEquals(LANGUAGE_MULTI, TranscribeOpts().language)
        assertTrue(isMultilanguage(TranscribeOpts().language))
    }

    @Test
    fun `no locale is ever inferred as a default`() {
        // Guards the exact regression: TranscribeOpts once defaulted to "de",
        // so every call site that omitted the argument silently pinned German.
        val default = TranscribeOpts().language
        assertFalse(
            default.equals("de", ignoreCase = true) || default.startsWith("de-", ignoreCase = true),
            "default language must not be a locale, got '$default'",
        )
    }

    @Test
    fun `an explicit pin survives`() {
        assertEquals("de", TranscribeOpts(language = "de").language)
        assertFalse(isMultilanguage("de"))
    }

    @Test
    fun `blank and auto count as multilanguage`() {
        assertTrue(isMultilanguage(null))
        assertTrue(isMultilanguage(""))
        assertTrue(isMultilanguage("  "))
        assertTrue(isMultilanguage("auto"))
        assertTrue(isMultilanguage("MULTI"))
    }

    /**
     * Per-provider translation, not one global string. The Android platform
     * recognizer expresses multilanguage by not pinning a tag at all; passing
     * the literal sentinel through would set an invalid BCP-47 value.
     */
    @Test
    fun `the platform recognizer expresses multilanguage as no language tag`() {
        assertNull(SystemSpeechRecognizerSession.toLanguageTag(LANGUAGE_MULTI))
        assertNull(SystemSpeechRecognizerSession.toLanguageTag(null))
        assertNull(SystemSpeechRecognizerSession.toLanguageTag("auto"))
        assertNull(SystemSpeechRecognizerSession.toLanguageTag(""))
    }

    @Test
    fun `the platform recognizer still expands an explicit pin`() {
        assertEquals("de-DE", SystemSpeechRecognizerSession.toLanguageTag("de"))
        assertEquals("en-US", SystemSpeechRecognizerSession.toLanguageTag("en"))
        assertEquals("fr-CA", SystemSpeechRecognizerSession.toLanguageTag("fr-CA"))
        assertEquals("pt-BR", SystemSpeechRecognizerSession.toLanguageTag("pt_BR"))
    }
}
